Fixed CORE-5611 - Higher memory consumption for prepared statements.

asfernandes opened the pull request:
- Remove some fields from nodes, including the pool reference and list of children nodes
- Organize some fields for better alignment and possible reduction on node size
- Delete DSQL scratch pool for DML statements

asfernandes committed:
Remove dsqlCompatDialectVerb field and organize some members in AggNode for better alignment.

asfernandes committed:
Remove kind field.

asfernandes committed:
Delete the scratch pool in the end of DML compilation.

asfernandes committed:
Process nodes using the scratch pool as default pool.

To make it work, change MAKE_parameter to allocate parameters in the message pool (statement pool).

asfernandes committed:
Remove pool reference and children lists in each node.

Children lists will be created and destroyed on demand.

asfernandes committed:
Avoid calls to JRD_get_thread_data.

asfernandes commented:
Can someone please review?

asfernandes commented:
Anyone to review? @dyemanov should all these changes later be backported to 3.0?

The commit fa821815fe808052ea58eea75fd5cdd268b473fc is an one that should be rewriten from scratch to backport.

dyemanov commented:
I'm reviewing it right now.

dyemanov commented on src/dsql/dsql.cpp:
```diff
@@ -1405,22 +1412,25 @@ static dsql_req* prepareStatement(thread_db* tdbb, dsql_dbb* database, jrd_tra*
+		Parser* parser = FB_NEW_POOL(*scratchPool) Parser(tdbb, *scratchPool, scratch, clientDialect,
 			scratch->getAttachment()->dbb_db_SQL_dialect, text, textLength,
 			tdbb->getAttachment()->att_charset);
 
```
What is the point in allocating the parser instance dynamically?

dyemanov commented on src/jrd/Attachment.cpp:
```diff
@@ -111,10 +111,11 @@ void Jrd::Attachment::destroy(Attachment* const attachment)
+		att_pools.add(pool);
 	return pool;
 }
 
```
Why is the parent-based new pool not accounted in att_pools? It's still a proper pool and must be destroyed via deletePool() in ~Attachment() if not deleted explicitly.

asfernandes replied:
When the scratch pool is manually deleted (DML statement compiled without error) the thing works with or without registering it in att_pools, as deletePool will be called and remove it.

But for the non-DML case or DML with exception in compilation, that means only the statement pool is explicitly destroyed (scratch is child of statement). In this case, it will try to destroy an already deallocated pool in detach.

Is there any real problem in deleting a child pool via its parent?

asfernandes replied:
After some tests I see that current usage (let child pool die with parent) does not decrement child pool's memory statistics when parent dies.

I think parent pools should maintain a list of its active children and when parent is destroyed it should decrement statistics from children.

AlexPeshkoff replied:
@asfernandes Does child pool unmap memory got from OS when parent is destroyed? As far as I remember we always used to explicitly destroy child pools before parent.

dyemanov replied:
> Is there any real problem in deleting a child pool via its parent?
AFAIK, it just never happened in our code, so I cannot answer for sure. "Parent" concept was used to redirect allocations up to some amount, but the child pool may have its own memory blocks that the parent is not aware of (and thus cannot deallocate).

asfernandes replied:
@AlexPeshkoff if I manually destroy parent then follow and destroy child, a segfault happens and this seems very correct, as should be the concept of child pool.

But if I only destroy parent, child does not decrement its usage in the memory stats object.

So IMO parent should do this for its active children when parent is deleted.

asfernandes replied:
When a child pool is going to allocate large amounts of memory, are these blocks requested from parent or are mapped directly with the OS?

dyemanov replied:
> But if I only destroy parent, child does not decrement its usage in the memory stats object.

AFAIU, the idea is that the parent should never die before any of its children. Once they're dead, it makes no sense to adjust their counters.

dyemanov replied:
In other words, "delete by pool" concept does not apply to child pools. And I'm not sure it should - it would complicate things without any real gain.

asfernandes replied:
Then how should you prefer? Making the scratch pool independent (not as child of statement) or use parent-child implementing some logic related to att_pools to never delete child after parent?

AlexPeshkoff replied:
> When a child pool is going to allocate large amounts of memory, are these blocks requested from parent or are mapped directly with the OS?

They are mapped directly from OS.

dyemanov replied:
I'd prefer to make the scratch pool independent but please ensure that it dies when necessary (prepare failure?).

asfernandes replied:
@dyemanov is it ok to merge in master now?

asfernandes committed:
Revert to statically alloc the parser instance.

Added a scope block to delete it before the scratch pool is destroyed in DsqlDmlRequest::dsqlPass.

asfernandes committed:
Better handling for the scratch pool and fixed CORE-5646.

CORE-5646 - Parse error when compiling a statement causes memory leak until attachment is disconnected.

asfernandes commented:
@dyemanov is it ok to merge in master now?

dyemanov commented:
I don't have other complaints, so it can be merged. Let's wait a bit before considering a backport into v3 though.

asfernandes committed:
Correction - thanks to Dmitry.

asfernandes committed:
Misc.

asfernandes committed:
Adjust after changes in master.

dyemanov commented:
@asfernandes , can we get back to this issue and backport into v3?

asfernandes commented:
> @asfernandes , can we get back to this issue and backport into v3?

I'm rebasing and completing the old patch.
